Ide refactoring inline method

SaarYogev opened the pull request:
Rebased then fixed compilation issues with #3176 .
I had the exact compilation error with `suggestedNames()` as the CI server and was able to fix it, @sohich would be great if you could take a look at this

SaarYogev commented:
Two more things:
1. I wasn't able to inline method with `:clion:runIde`. Might be because of a code piece that was too complicated, I'll try it later with a simpler example and update.

2. @sohich : 
> unit-tests and some improvements with type parameters substitution remain

Can you provide additional details, so other contributors and I could implement this?

SaarYogev commented:
Update:
Manual checks through `buildPlugin` went well, though seems like the option to remove the method after inline isn't implemented, it just stays there. Doesn't seem like a hard fix though ☺️ . Can someone give additional directions? Or accept the merge 😄

sirgl commented:
Notice, that it is not working with generics so far, so they will remain in code not substituted.
AFAIK, there is no ready way to substitute types outside of type inference, probably @vlad20012 may help you with it.
Also, there are no tests :) you can see, how they are implemented for another refactorings, e.g. [extract function](https://github.com/intellij-rust/intellij-rust/blob/master/src/test/kotlin/org/rust/ide/refactoring/RsExtractFunctionTest.kt)

SaarYogev commented:
Working on it for a few hours now, I got the first test ready, and added option for `removeDefinition` of the original fuction. One thing, could `{\n}\n;\n` around the inlined code be the intended behaviour? aka.
if the original code is:
```Kotlin
fn main() {
    test();
}
fn te/*caret*/st() {
    println!("test");
    println!("test2");
}
```
then what I expected was:
```Kotlin
fn main() {
    println!("test");
    println!("test2");
}
```
but what I got was:
```Kotlin
fn main() {
    {
        println!("test");
        println!("test2");
    }
    ;
}
```

SaarYogev commented:
No one answer, I'll continue assuming the `{\n}\n;\n` is redundant.
Unfortunately, I was unable to get rid of the empty statement (`;` alone), as it is **not present in the PSI tree!**
right now I'll continue writing the tests, if someone more experienced can confirm/deny this being this being an issue with the PSI that would be great!

ortem commented:
@SaarYogev Sorry for the long reply,

You're right, `{\n}\n;\n` is really redundant.

As for your problem with empty statement, you can explore [the Kotlin's implementation](https://github.com/JetBrains/kotlin/tree/master/idea/src/org/jetbrains/kotlin/idea/refactoring/inline) to find a proper way to replace old PSI with new and to perform such refactoring in general.

Note that `Inline function` should work in these cases too:
```Rust
fn foo() -> i32 { 42 }

fn main() {
    for i in 0..foo() {}
}
```

```Rust
fn foo() -> i32 { 42 }

fn main() {
    if foo() > 5 {}
}

```

SaarYogev commented:
Thanks for the comment @ortem , I added the tests you've suggested and they're indeed failing. the current reason is because it can't find a statement to repalce, as every ancestor is an expression (`RsInlineMethodProcessor.inlineWithLetBindingsAdded:129`). This time it doesn't seem to me like a PSI problem, just need to make it able to inline into expressions too ☺️ .

SaarYogev commented:
Well, apparently that was a 10 second fix, targeting expressions too if no statements were found did the trick 😅 .

SaarYogev commented:
Now finished copying all relevant tests from `RsExtractFunctionTest`. Most tests pass, with two problems remaining. 
First is the extra `;` problem, I tried using the `.bnf.preview` window but it's completely bugged for me, so I can't reallistacally fix problem with the PSI now. All tests which threw an extra `;` in a single place got their assertions adjusted accordingly. 
The Second is with tests of the `&self` parameter. my test doesn't reverse exactly to `RsExtractFunctionTest`'s result, but here I'm if my result is actually wrong.

Original test:
```Kotlin
fun `test extract a function with the parameter ref self`() = doTest("""
            struct S;
            impl S {
                fn foo(&self) {
                    <selection>println!("test");
                    println!("test2");
                    self.test();</selection>
                }

                fn test(&self) {
                    println!("bla");
                }
            }
        """, """
            struct S;
            impl S {
                fn foo(&self) {
                    self.bar();
                }

                fn bar(&self) {
                    println!("test");
                    println!("test2");
                    self.test();
                }

                fn test(&self) {
                    println!("bla");
                }
            }
        """,
        false,
        "bar")
```
My Test, changed to the current result instead of supposed assertion:
```Kotlin
fun `test inline impl method with ref self parameter`() = doTest("""
            struct S;
            impl S {
                fn foo(&self) {
                    self.bar();
                }

                fn b/*caret*/ar(&self) {
                    println!("test");
                    println!("test2");
                    self.test();
                }

                fn test(&self) {
                    println!("bla");
                }
            }
            """, """
            struct S;
            impl S {
                fn foo(&self) {
                    println!("test");
                    println!("test2");
                    &self.test();
                    ;
                }

                fn test(&self) {
                    println!("bla");
                }
            }
            """)
```
Notice it changed `self.test();` to `&self.test();` - that's the difference.

Beside those I have 2 tests(`test inline function with expression return value`, `test inline function with return value and no parameters, with range operator and assignment in function`), which seem to not pass because of more ordinary bugs, I'll have a look at them soon, but they worry me less.

SaarYogev commented:
Finally fixed most errors, including a workarounds for the extra `;`  - as I explained, I am currently unable to debug `.bnf` files, so I just manually removed them through code.
The only thing that remains is the ref/mut issue that I brought up earlier - @ortem just decide what's correct and I'll fix the tests accordingly.

SaarYogev commented:
@vlad20012 @ortem @sirgl @sohich This needs review 😅

SaarYogev commented:
I still need an answer for *&self* test (full example [in this comment](https://github.com/intellij-rust/intellij-rust/pull/4460#issuecomment-541353706)) @ortem @vlad20012 .
And I'd like a review, I'll clean up so it'll be easier.

SaarYogev commented:
Rebased again. stable build passes now, nightly fails because it doesn't `'all.invocations.keep.the.method` in PropertyResouceBundle (part of ResouceBundle). Couldn't find any docs on this, does anyone have an idea?

SaarYogev commented:
Well kind of fixed the `RefactoringBundle` issue. You see, `"all.invocations.keep.the.method"` was moved [here](https://github.com/JetBrains/intellij-community/commit/25133895679652029f783979d91a68e7214387f7) from `RefactoringBundle` to the `JavaRefactoringBundle`. It's not sensible for us to reference the `JavaRefactoringBundle` in *rust*. But right now it a hardcoded *string*, meaning locales won't work for this :X. The real fix is to PR *IntelliJ community*.

SaarYogev commented:
@ortem / @vlad20012 Could you CR/merge this, now that build is fixed?

SaarYogev commented:
Got an answer about `JavaRefactoringBundle` from @chashnikov in [IDEA-237145](https://youtrack.jetbrains.com/issue/IDEA-237145). Terms like *method* are different between languages, so it was moved to a language-specific refactoring bundles. From what I understood, we should make a `RustBundle` and start putting language specific terms there.
If you agree say it (:+1:) and I'll open a separate issue on that.

Also, I checked and *Rust* uses the term *function*, so I'll change the code accordingly.

chashnikov commented:
As far as I can see Rust sources aren't internationalized, UI-visible strings are hardcoded in sources. So for now you can just put the string to the sources as well.

SaarYogev commented:
I see, so for now hardcoded strings are fine, and we'll add a `RustBundle` when we'll need i18n support. Thanks!

SaarYogev commented:
@ortem @vlad20012 This is ready and waiting for CR/merge, awaiting your help.

SaarYogev commented on CONTRIBUTORS.txt:
```diff
@@ -120,3 +120,5 @@ webdesus
 winger
 zinoviy23
 zjhmale
+gejun123456
```
Bad rebase on my part, he already appears on line 41. Also moved myself to the correct alphabetic place now.

vlad20012 commented:
Soo could you please squash all these commits into 1?

SaarYogev commented:
Yeah working on it 😅

SaarYogev commented:
@vlad20012 Done, and fixed your comments. 
Just didn't fix error handling, I'll need examples to make it better, I just fixed the critical stuff, like removing `!!` or not just absorbing errors.

vlad20012 commented:
Ok, I think I'll merge in on the next week (after the next release branch creation).
I'll can fix error handling by myself after the merge. 
But if you're really interested, take a look at Java's `InlineMethodProcessor.preprocessUsages`. Note that `performRefactoring` method really must not throw exceptions.

SaarYogev commented:
Got an odd failure on an unrelated test - `RsExplainFilterTest.test new explain format`, just on a specific combination of OS, rust and IDE versions. Besides that build passes. Gonna write some tests for `preprocessUsages` now.

SaarYogev commented:
@vlad20012 I saw preprocess is for the usages, what should I do if the method itself can't be inlined? For now I just caught all errors inside `performRefactoring`, but I assume there's something better. Also, there some earlier code that throwed inside `findUsages` not sure if that's right either.

vlad20012 commented:
> what should I do if the method itself can't be inlined

You can check the function before performing inlining in `RsInlineFunctionHandler.inlineElement` and show errors there. See Java's `InlineMethodHandler.performInline`

SaarYogev commented:
@vlad20012 Think I'm done.
Now every exceptional case is checked in either `inlineElement()` or `preprocessUsages()`. 
I still throw inside performRefactoring if I checked the case earlier but the compiler doesn't know what to do, but I catch it anyways in `catch (e: IncorrectOperationException)`.
I tried to make exception throwing more explicit, if you think I should just remove it and let them throw random NPEs and such then `catch (e: Exception)` I can do that too.

vlad20012 commented:
Well, let's merge it
bors r+

Undin commented:
@sohich @SaarYogev Thanks!

Kobzol commented:
This refactoring should also replace the function parameters with the actual called arguments, is that right? At least that's what Kotlin plugin does for function inlining.

```rust
fn foo<T>(a: u32, b: &T) -> u32 {
    a
}

fn main() {
    let a = foo(1, "hello");
}

vvv

fn main() {
    let a = a; // should be let a = 1;
}
```

Was there some issue with that or is it not implemented yet?

vlad20012 commented:
@Kobzol I'm going to open an issue with a list of further improvements of the refactoring
